docs/architecture-decisions/flag-type.md

```diff
@@ -0,0 +1,133 @@
+
+The root cause of this issue is the absence of a dedicated, authoritative type definition for the flag itself. Instead, the type is inferred from the `defaultVariant` or API itself (`T` from `Resolve<T>`) , which is not always a reliable source of truth for all possible variants. This can lead to unexpected errors and make it difficult for developers to debug their feature flags.
+
+
+## Requirements
+
+* The new `type` field in the flag definition must be optional to ensure backward compatibility.
+* If the `type` field is present, `flagd` must validate that all variants of the flag conform to this type during initialization.
+* Type mismatches found during initialization must be reported as errors.
+* The `Resolve<T>` and `ResolveAll` methods must use the `type` field for validation when it is available.
+* The implementation must be consistent with the OpenFeature specification and the flag manifest schema.
+
+
+## Considered Options
+
+* **Consistent `defaultVariant` Validation:** Align the behavior of `Resolve<T>` with `ResolveAll` by making `Resolve<T>` validate the evaluated variant against the type of the `defaultVariant`.
+* **API Extension with Explicit Flag Type:** Introduce an optional `type` property to the flag definition to serve as the authoritative source for type validation.
+
+
+## Proposal
+
+This proposal is to extend the flag definition with an optional `type` property. This approach is chosen over simply aligning the `Resolve<T>` and `ResolveAll` validation because it addresses the root cause of the type inconsistency and provides a more robust, long-term solution.
+
+By introducing an explicit `type` field, it establishes a single source of truth for the flag's type, independent of its variants. This allows for early and consistent type validation during flag definition parsing, preventing type-related errors at runtime.
+
+The new `type` field will be optional to maintain backward compatibility with existing flag configurations. If the field is omitted, `flagd` will treat the flag as having `object`, and no type validation will be performed against the `defaultVariant`. When the `type` field is present, `flagd` will enforce that all variants of the flag conform to the specified type.
+
+This change will make the behavior of `flagd` more predictable and reliable.
+
+
+### API changes
+
+The `flagd` flag definition will be updated to include an optional `type` property. This property will be a string enum with the following possible values: `"boolean"`, `"string"`, `"number"`, and `"object"`.
```

Comment:
The OpenFeature CLI defines `float`, `integer`, `object`, `string`, and `boolean`. We should probably keep them in sync but I'd like to get others thoughts on this.